components/ai_chat/core/browser/conversation_handler.cc

```diff
@@ -0,0 +1,1385 @@
+                       weak_ptr_factory_.GetWeakPtr()));
+  } else {
+    // Run immediately if already initialized
+    text_embedder_->GetTopSimilarityWithPromptTilContextLimit(
+        prompt, text, context_limit, std::move(callback));
+  }
+}
+
+void AssociatedContentDelegate::OnTextEmbedderInitialized(bool initialized) {
+  if (!initialized) {
+    VLOG(1) << "Failed to initialize TextEmbedder";
+    for (auto& callback_info : pending_top_similarity_requests_) {
+      std::move(std::get<3>(callback_info))
+          .Run(base::unexpected<std::string>(
+              "Failed to initialize TextEmbedder"));
+    }
+    pending_top_similarity_requests_.clear();
+    return;
+  }
+
+  CHECK(text_embedder_);
+  for (auto& callback_info : pending_top_similarity_requests_) {
+    text_embedder_->GetTopSimilarityWithPromptTilContextLimit(
+        std::get<0>(callback_info), std::get<1>(callback_info),
+        std::get<2>(callback_info), std::move(std::get<3>(callback_info)));
+  }
+  pending_top_similarity_requests_.clear();
+}
+
+ConversationHandler::ConversationHandler(
+    const mojom::Conversation* conversation,
+    AIChatService* ai_chat_service,
+    ModelService* model_service,
+    AIChatCredentialManager* credential_manager,
+    AIChatFeedbackAPI* feedback_api,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
+    : metadata_(conversation),
+      ai_chat_service_(ai_chat_service),
+      model_service_(model_service),
+      credential_manager_(credential_manager),
+      feedback_api_(feedback_api),
+      url_loader_factory_(url_loader_factory) {
+  // When a client disconnects, let observers know
+  receivers_.set_disconnect_handler(
+      base::BindRepeating(&ConversationHandler::OnClientConnectionChanged,
+                          weak_ptr_factory_.GetWeakPtr()));
+  models_observer_.Observe(model_service_.get());
+  // TODO(petemill): differ based on premium status, if different
+  ChangeModel(model_service->GetDefaultModelKey());
+}
+
+ConversationHandler::~ConversationHandler() {
+  if (associated_content_delegate_) {
+    associated_content_delegate_->OnRelatedConversationDestroyed(this);
+  }
+}
+
+void ConversationHandler::AddObserver(Observer* observer) {
+  observers_.AddObserver(observer);
+}
+
+void ConversationHandler::RemoveObserver(Observer* observer) {
+  observers_.RemoveObserver(observer);
+}
+
+void ConversationHandler::Bind(
+    mojo::PendingReceiver<mojom::ConversationHandler> receiver,
+    mojo::PendingRemote<mojom::ConversationUI> conversation_ui_handler) {
+  receivers_.Add(this, std::move(receiver));
+  conversation_ui_handlers_.Add(std::move(conversation_ui_handler));
+  OnClientConnectionChanged();
+  // In some cases, this page handler hasn't been created and remote might not
+  // have been set yet.
+  // ex. A user may ask a question from the location bar
+  if (!pending_conversation_entry_.is_null()) {
+    OnHistoryUpdate();
+  }
```

Comment:
I don't quite understand this, what's the relationship between whether pending_conversation_entry is null and if page handler is there and remote has been set?
Also, why is OnHistoryUpdate called when binding conversation?

Reply:
This was copied from the existing client connection code: https://github.com/brave/brave-core/blob/494f3a12db100e8c90c116be48ea08d47990a177/browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc#L108-L114

Doesn't make sense to me either - will remove and see if there's an issue with pending entry from the omnibox